### PR TITLE
Additional verification for VaRA in buildsetup

### PR DESCRIPTION
**Pull request by LuAbelt**
Added additional verification to the `verify_install` step for VaRA

Additionally, VaRA now runs the `check_vara` target to verify that it not only compiles but also passes unittests.

Not sure if we really want this in a separate `verify_build` but in `verify_install` we currently don't have access to the build directory in which we run `ninja`.

resolves https://github.com/se-sic/VaRA/issues/801

### CODE REVIEW

File: varats/varats/tools/research_tools/research_tool.py [hunk 1]
```diff
@@ -490,6 +490,17 @@ def verify_install(self, install_location: Path) -> bool:
...
+    def verify_build(
+        self, build_type: BuildType, build_folder_suffix: tp.Optional[str]
+    ) -> bool:
+        """
+        Verify if the research tool was built correctly for a given build_type.
+
+        Returns:
+            True, if the build was correct.
+        """
+        return True
```
Comment (vulder) [inline varats/varats/tools/research_tools/research_tool.py]:
```suggestion
```

Should we really provide a default impl here? Could be better to force every `research_tool` to implement the method by making it abstract. If the tool wants to default to `true` they can.

File: varats/varats/tools/research_tools/research_tool.py [hunk 2]
```diff
@@ -490,6 +490,17 @@ def verify_install(self, install_location: Path) -> bool:
             True, if the tool was correctly installed
         """
 
+    def verify_build(
+        self, build_type: BuildType, build_folder_suffix: tp.Optional[str]
+    ) -> bool:
+        """
+        Verify if the research tool was built correctly for a given build_type.
+
```
Comment (vulder) [inline varats/varats/tools/research_tools/research_tool.py:499]:
```suggestion

            build_type: which type of build should be used, e.g., debug,
                        development or release
            build_folder_suffix: a suffix that is appended to the build folder
```

PS: not sure if this is correctly aligned, gh makes it hard so see this

File: varats/varats/tools/research_tools/vara.py [hunk 3]
```diff
@@ -374,8 +374,58 @@ def verify_install(self, install_location: Path) -> bool:
         status_ok &= (install_location / "bin/opt").exists()
         status_ok &= (install_location / "bin/phasar-llvm").exists()
 
+        # Check that clang++ can display it's version
+
+        clang = local[str(install_location / "bin/clang++")]
+
+        (ret, stdout, _) = clang.run("--version")
```
Comment (vulder) [inline varats/varats/tools/research_tools/vara.py]:
```suggestion
        # Check that clang++ can display it's version
        clang = local[str(install_location / "bin/clang++")]
        ret, stdout, _ = clang.run("--version")
```
Python does not require braces to unpack the return values

File: varats/varats/tools/research_tools/vara.py [hunk 4]
```diff
@@ -374,8 +374,58 @@ def verify_install(self, install_location: Path) -> bool:
...
+
+        vara_name = self.code_base.get_sub_project("vara-llvm-project").name
+        status_ok &= (ret == 0)
```
Comment (vulder) [inline varats/varats/tools/research_tools/vara.py]:
```suggestion
        status_ok &= (ret == 0)
```
What's the idea behind putting braces around the comparision?

File: varats/varats/tools/research_tools/vara.py [hunk 5]
```diff
@@ -374,8 +374,58 @@ def verify_install(self, install_location: Path) -> bool:
...
+
+        status_ok &= (vara_name in stdout)
+
+        # Check that phasar-llvm can display it's version
+        phasar_llvm = local[str(install_location / "bin/phasar-llvm")]
+
+        (ret, stdout, _) = phasar_llvm.run("--version")
+
+        status_ok &= (ret == 0)
```
Comment (vulder) [inline varats/varats/tools/research_tools/vara.py]:
```suggestion
        phasar_llvm = local[str(install_location / "bin/phasar-llvm")]
        ret, stdout, _ = phasar_llvm.run("--version")
        status_ok &= ret == 0
```

File: varats/varats/tools/research_tools/vara.py [hunk 6]
```diff
@@ -374,8 +374,58 @@ def verify_install(self, install_location: Path) -> bool:
...
 
+    def verify_build(
+        self, build_type: BuildType, build_folder_suffix: tp.Optional[str]
+    ) -> bool:
+        """
+        Verifies whether vara was built correctly for the given target.
+
+        Args:
+            build_type: which type of build should be used, e.g., debug,
+                        development or release
```
Comment (vulder) [inline varats/varats/tools/research_tools/vara.py:403]:
```suggestion
            build_type: which type of build should be used, e.g., debug,
                        development or release
            build_folder_suffix: a suffix that is appended to the build folder
```

File: varats/varats/tools/research_tools/vara.py [hunk 7]
```diff
@@ -374,8 +374,58 @@ def verify_install(self, install_location: Path) -> bool:
...
+            return False
+
+        build_folder_path = build_type.build_folder(build_folder_suffix)
+        full_path /= build_folder_path
+
+        ninja = local["ninja"].with_cwd(full_path)
+
+        (ret, _, _) = ninja.run("check-vara")
+
+        return bool(ret == 0)
```
Comment (vulder) [inline varats/varats/tools/research_tools/vara.py]:
```suggestion
        ninja = local["ninja"].with_cwd(full_path)
        ret, _, _ = ninja.run("check-vara")
        return ret == 0
```

**Review by vulder — CHANGES_REQUESTED** [review]
Impl looks good, thx.

I only have a few small issue/questions.

File: varats/varats/tools/research_tools/vara.py (hunk 7 above)
Comment (LuAbelt) [reply varats/varats/tools/research_tools/vara.py]: If we remove the explicit cast to `bool`, mypy will complain about the function returning `Any` even though it was declared as `bool`

**Review by boehmseb — APPROVED** [review]

**Review by vulder — APPROVED** [review]

**Pull request merged**